### ISSUE TITLE
AEAD Algorithm AES_SIV_CMAC_256

### ISSUE THREAD

**Issue by aaronbojarski** [beevik/nts#1]
Hi Brett Vickers

I hope you don't mind me reaching out here.

Great to see that you are working on a NTS implementation in Go.
We are also working on NTS in Go, although slightly modified for the
SCION internet architecture.
https://github.com/marcfrei/scion-time

First I wanted to let you know some insights we gained about
AES_SIV_CMAC_256.
As you probably have noticed it is not supported by Go Crypto or Googles
Tink.
We therefore also started by using SIV-GO. It however has an issue when
using the assembly instructions on x86 machines. There already exists an
issue for the error that we also encountered. There does not seem to be
any more development that would fix the issue.

We then went looking for alternatives and found the archived
https://github.com/miscreant/miscreant.go
It does not have the same issue and it is more efficient (tested on ARM
machines).

Do you have any thoughts on SIV-GO or plans to implement
AES_SIV_CMAC_256 yourself?

Thank you and best regards
Aaron

**Comment by beevik** [comment]
I ran into an issue with siv-go too.  The assembly implementation was written using amd64 instructions that expected memory to be aligned.  I modified my code to always work with aligned data and haven't had any problems since.

**Comment by aaronbojarski** [comment]
Thank you for your response. 
True, we could have also aligned the data. I might look at that again. I guess I just did not want to work with siv-go anymore once it threw an error that is already known.

Looking forward to see where your development goes.

**Comment by beevik** [comment]
Sadly there do not seem to be any maintained go implementations of AES_SIV_CMAC.  I believe I saw something recently on one of the go issue trackers that they may be working on a standard library version, though.

**Comment by aaronbojarski** [comment]
Thank you. Good to know.

**Comment by mdavids** [comment]
FWIW: I ran into similar issues with [secure-io/siv-go](https://github.com/secure-io/siv-go) myself, even with the [code changes](https://github.com/beevik/nts/blob/444360a81efa36c53f63b3b13d3dcb556db2a1c4/session.go#L29) Brett made, and was able to work around them by patching [newGCM()](https://github.com/secure-io/siv-go/blob/5ff40651e2c4476835bf9f18a6bee513f588d9bc/aes_gcm_amd64.go#L21) in [aes_gcm_amd64.go](https://github.com/secure-io/siv-go/blob/master/aes_gcm_amd64.go) like this:

```
import (
//      "crypto/aes"
        "crypto/cipher"
        "crypto/subtle"

//      "golang.org/x/sys/cpu"
)
```

and

```
        func newGCM(key []byte) aead {
        //      if cpu.X86.HasAES && cpu.X86.HasPCLMULQDQ {
        //              block, _ := aes.NewCipher(key)
        //              return &aesGcmSivAsm{block: block, keyLen: len(key)}
        //      }
                return newGCMGeneric(key)
        } 
```

Steps:

- git clone https://github.com/secure-io/siv-go.git /tmp/siv-go-debug
- cd /tmp/siv-go-debug
- git checkout 5ff40651e2c4

In `go.mod` I did:

`replace github.com/secure-io/siv-go => /tmp/siv-go-debug`

For my use case, losing some hardware acceleration is not a problem at all, and the annoying crashes have stopped.

There is also `aes_gcm_noasm.go` by the way that could have been tweaked a bit. But I choose this hack over changing any constraints.

**Comment by beevik** [comment]
Marco, I think your patch works only because it totally disables hardware AES. I would prefer not to do this if possible. Can you give any more details on the errors you were seeing without your change? Because I have not enountered these errors.  

Useful information you could provide:
* The machine architecture you are using (CPU).
* The operating system.
* A test scenario that reproduces the bug.

Thanks!

**Comment by mdavids** [comment]
Hi Brett,

I don't consider it a proper patch, but merely a 'hacky' workaround.  But for me it worked, so I figured to document it here for the benefit of others. Indeed is disables hardware AES.

I encountered this on my Ubuntu 24.04.4 LTS VM's - kernel version `6.8.0-124-generic` (on a VMWare host).

```
uname -m
x86_64
```

```
cat /proc/cpuinfo | grep 'model name' | tail -1
model name	: Intel(R) Xeon(R) Gold 5115 CPU @ 2.40GHz
```

Running this code: https://github.com/mdavids/ntptools/tree/main/go/ntp-exporter-multitarget

Logging attached:

[crashes.txt](https://github.com/user-attachments/files/29954431/crashes.txt)

**Comment by beevik** [comment]
Do you think you could do me a favor and test the NTS package by replacing `session.go` line 320 with this:

```go
plaintext, err := s.cipherS2C.Open(nil, nonce, ciphertext, align(buf[:offset]))
```

(This adds the `align()` call around `buf[:offset]`).

**Comment by mdavids** [comment]
I was about to try this, but there is one problem: I can no longer reproduce the crashes. I’m not sure why, since I’m following exactly the same steps as before. I’ll look into why I can no longer reproduce the crashes and keep you posted.

However; according to my logs, I did already try this exact change at line 320. Unfortunately, it did not fix the problem at that time.

I’ll give it another try once I’m able to reproduce the crashes again 🙂.

**Comment by beevik** [comment]
I believe the cause of this bug has been identified by another contributor. The fix is available in release [v0.3.3](https://github.com/beevik/nts/tree/v0.3.3).
